capsules/src/usb.rs

```diff
@@ -0,0 +1,467 @@
+    InterfacePower,
+}
+
+fn get_descriptor_type(byte: u8) -> Option<DescriptorType> {
+    match byte {
+        1 => Some(DescriptorType::Device),
+        2 => Some(DescriptorType::Configuration),
+        3 => Some(DescriptorType::String),
+        4 => Some(DescriptorType::Interface),
+        5 => Some(DescriptorType::Endpoint),
+        6 => Some(DescriptorType::DeviceQualifier),
+        7 => Some(DescriptorType::OtherSpeedConfiguration),
+        8 => Some(DescriptorType::InterfacePower),
+        _ => None,
+    }
+}
+
+/// Get a descriptor type that is legal in a SetDescriptor request
+fn get_set_descriptor_type(byte: u8) -> Option<DescriptorType> {
+    match get_descriptor_type(byte) {
+        dt @ Some(DescriptorType::Device) => dt,
+        dt @ Some(DescriptorType::Configuration) => dt,
+        dt @ Some(DescriptorType::String) => dt,
+        _ => None
+    }
+}
+
+#[derive(Copy, Clone)]
+pub struct DeviceRequestType(u8);
+
+impl DeviceRequestType {
+    pub fn transfer_direction(self) -> TransferDirection {
+        match self.0 & (1 << 7) {
+            0 => TransferDirection::HostToDevice,
+            _ => TransferDirection::DeviceToHost,
+        }
+    }
+
+    pub fn request_type(self) -> RequestType {
+        match (self.0 & (0b11 << 5)) >> 5 {
+            0 => RequestType::Standard,
+            1 => RequestType::Class,
+            2 => RequestType::Vendor,
+            _ => RequestType::Reserved,
+        }
+    }
+
+    pub fn recipient(self) -> Recipient {
+        match self.0 & 0b11111 {
+            0 => Recipient::Device,
+            1 => Recipient::Interface,
+            2 => Recipient::Endpoint,
+            3 => Recipient::Other,
+            _ => Recipient::Reserved,
+        }
+    }
+}
+
+impl fmt::Debug for DeviceRequestType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{{{:?}, {:?}, {:?}}}",
+               self.transfer_direction(), self.request_type(), self.recipient())
+    }
+}
+
+#[derive(Debug)]
+pub enum TransferDirection {
+    DeviceToHost,
+    HostToDevice,
+}
+
+#[derive(Debug)]
+pub enum RequestType {
+    Standard,
+    Class,
+    Vendor,
+    Reserved,
+}
+
+#[derive(Debug)]
+pub enum Recipient {
+    Device,
+    Interface,
+    Endpoint,
+    Other,
+    Reserved,
+}
+
+#[derive(Debug)]
+pub enum FeatureSelector {
+    DeviceRemoteWakeup,
+    EndpointHalt,
+    TestMode,
+    Unknown,
+}
+
+impl FeatureSelector {
+    fn get(value: u16) -> Self {
+        match value {
+            1 => FeatureSelector::DeviceRemoteWakeup,
+            0 => FeatureSelector::EndpointHalt,
+            2 => FeatureSelector::TestMode,
+            _ => FeatureSelector::Unknown,
+        }
+    }
+}
+
+pub trait Descriptor {
+    /// Serialized size of Descriptor
+    fn size(&self) -> usize;
+
+    /// Serialize the descriptor to a buffer for transmission on the bus
+    fn write_to(&self, buf: &mut [u8]) -> usize;
+}
+
+pub struct DeviceDescriptor {
```

Comment:
Comments in general for these structure would be good, but especially for this one and it's fields which are somewhat obscure if you don't know the USB protocol. Even just adapting the basic descriptions from the [USB Made Simple](http://www.usbmadesimple.co.uk/) would be enough I think:

![screenshot](https://cloud.githubusercontent.com/assets/4689/26756659/ce2abd34-4875-11e7-8001-99bcf0b8f7d2.png)